message TableConfiguration {

    // The relation name of this table.
    required string name = 1;

    // The primary key field name for this table.
    required string primary_key = 2;

    // The columns that will be monitored on this table.
    repeated string columns = 3;

    // The related tables that are part of this group. (Note that all joins
    // listed here will be `LEFT JOIN`s with respect to this table.)
    repeated JoinConfiguration joins = 4;

}

message JoinConfiguration {

    // The child (right) that is being joined to the parent (left) table.
    required TableConfiguration table = 1;

    // The foreign key column from the child (right) table to be joined on the
    // parent (left) table's primary key. (For instance, if this represented a
    // relationship between an `author` and `book` table, this field value
    // would correspond to `book.author_id`.)
    // XXX: For practicality's sake, only optional 1:{0,1} relationships (where
    // the child table has a `UNIQUE` constraint on the foreign key) are
    // supported.
    required string foreign_key = 2;

}

message DatabaseConfiguration {

    // A user-defined alias for this database. (This is is only used for
    // administrator convenience, and doesn't have any impact on the behavior
    // of any component.)
    required string name = 1;

    message ConnectionConfiguration {

        // http://www.postgresql.org/docs/current/static/libpq-connect.html#LIBPQ-CONNSTRING
        required string dsn = 1;

    }

    // The administrative connection for this database.
    required ConnectionConfiguration connection = 2;

}

message GroupConfiguration {

    // The database where this group resides.
    required DatabaseConfiguration database = 1;

    // The primary table to be monitored with this group.
    required TableConfiguration table = 2;

}
